Add blockAccessibilityFocus flag

hannah-hyj opened the pull request:
Add a new flag for a11y focusable

- Accessibility focus, which is the focus used by screen readers like TalkBack and VoiceOver, is different from input focus.
-  Our current logic use some existing flags to decide if a node is accessibilty focusable. like "if it's a slider / has a check state / has keyboard focus/..., then it's a11y focusable" https://github.com/flutter/flutter/blob/ecbb115ae3a8cba2977ecab9f52086df860cfb1a/engine/src/flutter/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L98
- but we lack the ability to explicitly set a node to be unfocusable in a11y! 
- This flag can be used to explicitly set some semantics nodes to be unfocusable in a11y mode. if it is false, we fall back to the logic "if it's a slider / has a check state / has keyboard focus/..., then it's a11y focusable" 

future use case 1:
     user can set live region to be not focusable, so when content changes, it will still announce, but the content can't be focused by swiping. 
future use case 2:
    when pushing a new route like a dialog, setting the semantics nodes in old pages to be un focusable. 




## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

hannah-hyj committed:
add comments for tests

1

resolve comments

rename config and properties flag to isSubTreeAccessibilityFocusable

resolve comments

Update packages/flutter/lib/src/semantics/semantics.dart

Update semantics_test.dart

lint

Update semantics_test.dart

1

6

5

5

Update embedder_semantics_update.cc

lint

add flag

4

3

Update semantics.dart

2

1

avoid abbreviation

avoid abbreviation

lint

hannah-hyj commented:
cc @ash2moon

chunhtai reviewed (commented):
Should the focusability applies the subtree? The current code seems to only disable the focusability of the node but not the sub tree.

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -1715,6 +1720,16 @@ class SemanticsProperties extends DiagnosticableTree {
   /// element it is reading, and is separate from input focus.
   final bool? focused;
 
+  /// If non-null, whether the subtree can be focused by accessibility services.
```
some where in the doc should mention the behavior when this is false but somewhere in the subtree set this to true. Will that part of the subtree because focusable or it will get override by the parent.

chunhtai reviewed (commented):
Still missing a test for the _semanticsGeneral check in 
flutter/packages/flutter/test/widgets/semantics_role_checks_test.dart

hannah-hyj committed:
blockAccessibilityFocus

lint

1

use blockAccessibilityFocus instead

Update AccessibilityBridge.java

Update semantics.dart

Update semantics.dart

hannah-hyj commented:
discussed offline and will add options to block a subtree or block a current node

hannah-hyj committed:
make BlockAccessibilityFocus an enum

fix tests

chunhtai commented on engine/src/flutter/lib/ui/semantics/semantics_flags.cc:
```diff
@@ -40,7 +40,8 @@ void NativeSemanticsFlags::initSemanticsFlags(
     bool isSlider,
-    bool isKeyboardKey) {
+    bool isKeyboardKey,
+    bool blockAccessibilityFocus) {
```
```suggestion
    bool isAccessibilityFocusBlocked) {
```

chunhtai reviewed (commented):
We also need to consider in the case of implicit merge, whether carry blockaccessibilityFocus up to the parent the right thing to do.

hannah-hyj committed:
add  a test for merge semantics

lint

lint

lint

lint

rename params

rename params

rename params

hannah-hyj commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -114,6 +114,22 @@ typedef SemanticsUpdateCallback = void Function(SemanticsUpdate update);
+///
+/// This is typically used to prevent screen readers
+/// from focusing on parts of the UI.
+enum BlockAccessibilityFocus {
```
thanks, the new name is way more clear

chunhtai commented on engine/src/flutter/lib/ui/semantics.dart:
```diff
@@ -1575,6 +1596,9 @@ class SemanticsFlags extends NativeFieldWrapperClass1 {
   bool hasRepeatedFlags(SemanticsFlags other) {
+    // isAccessibilityFocusBlocked is not checked in `hasRepeatedFlags` because
+    // two semantics nodes both with `isAccessibilityFocusBlocked` set to true is
+    // still compatible and not conflicting.
```
In this case, the method name no longer correct. Consider rename to something more descriptive

chunhtai replied:
we should create a new method and deprecate this one.

hannah-hyj committed:
Update semantics_test.dart

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -6421,6 +6462,10 @@ class SemanticsConfiguration {
     if (_flags.hasRepeatedFlags(other._flags)) {
       return false;
     }
+    if (_flags.isAccessibilityFocusBlocked != other._flags.isAccessibilityFocusBlocked) {
```
if one is BlockSubTree another one is BlockCurrent, I think they can merge to be BlockSubTree

hannah-hyj replied:
added to the absorb

hannah-hyj committed:
Update semantics.dart

hannah-hyj committed:
rename hasRepeatedFlags to hasConflictingFlags

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -114,6 +114,40 @@ typedef SemanticsUpdateCallback = void Function(SemanticsUpdate update);
+  /// may still be focusable.
+  blockNode;
+
+  /// The AccessiblityFocusBlockType when two node get merged.
```
```suggestion
  /// The AccessiblityFocusBlockType when two nodes get merged.
```

chunhtai reviewed (approved):
LGTM

hannah-hyj committed:
resolve comments

hannah-hyj committed:
Merge branch 'master' into add-a11y-focusable-flag

hannah-hyj committed:
lint

hannah-hyj committed:
typo

hannah-hyj committed:
Merge branch 'master' into add-a11y-focusable-flag
